compiler/src/dotty/tools/dotc/typer/Applications.scala

```diff
@@ -1427,7 +1427,17 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
         narrowByTypes(alts, args, resultType)
 
       case pt =>
-        alts filter (normalizedCompatible(_, pt))
+        val noSam = alts filter (normalizedCompatible(_, pt))
+        if (noSam.isEmpty) {
+          pt match {
+            case SAMType(mtp) =>
+              val sam = narrowByTypes(alts, mtp.paramInfos, mtp.resultType)
+              if (sam.nonEmpty && !pt.classSymbol.hasAnnotation(defn.FunctionalInterfaceAnnot))
+                ctx.warning(ex"$pt does not have the @FunctionalInterface annotation.", ctx.tree.pos)
```

Comment:
`ctx.tree.pos` here will not be very precise. Ideally it should be the position of the tree being expended:
```scala
-- Warning: tests/allan/Test.scala:5:6 -----------------------------------------
5 |  val out2 = new ObjectOutputStream(println)
  |  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Should be:
```scala
-- Warning: tests/allan/Test.scala:5:6 -----------------------------------------
5 |  val out2 = new ObjectOutputStream(println)
  |                                    ^^^^^^^
```
```